File: 0066.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        res = []
        digits.reverse()
        for i in range(len(digits)):
            if digits[i] == 9:
                if i < len(digits) - 1:
                    res.append(0)
                else:
                    res.extend([0, 1])
            else:
                digits[i] += 1
                res.extend(digits[i:])
                break
        res.reverse()
        return res


class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        num = ''
        for digit in digits:
            num += str(digit)
        num = int(num) + 1
        return [v for v in str(num)]

File: 0066.cpp
// First time
/*
    Optimize
    1. Return earlier because we only add one so at most 9 -> 10
    2. Insert in the end to avoid extra time complexity
    T: O(n)/S: O(1)
*/
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        for (int i = n - 1; i >= 0; --i) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        digits[0] = 1;
        digits.push_back(0);
        return digits;
    }
};

/*
    insert in the begin
    T: O(n)/S: O(1)
*/
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        bool carry = false;
        int n = digits.size();
        digits[n - 1]++;
        for (int i = n - 1; i >= 0; --i) {
            digits[i] += carry;
            carry = digits[i] / 10;
            digits[i] %= 10;
        }
        if (carry)
            digits.insert(digits.begin(), 1);
        /*
            digits[0] = 1;
            digits.push_back(0);
        */
        return digits;
    }
};

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        bool carry = false;
        int n = digits.size();
        for (int i = n - 1; i >= 0; --i) {
            int tmp = 0;
            if (i == n - 1) {
                tmp = digits[i] + 1;
            } else {
                tmp = carry + digits[i];
            }
            digits[i] = tmp % 10;
            carry = tmp / 10;
        }
        if (carry)
            digits.insert(digits.begin(), 1);
        return digits;
    }
};